### PR TITLE
adds support for ping/kill during maintenance stop/start

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- adds support for ping during maintenance stop
- adds support for service kill during maintenance start

## Why are we making these changes?

In addition to configuring maintenance mode, we would like to kill/ping running services for the token when enabling/disabling maintenance. The new behaviors are added as default options and can be disabled by CLI flags.

### help flag

```
$ waiter maintenance start -h
usage: waiter maintenance start [-h] [--ask-kill | --force-kill | --no-kill] token message
...
optional arguments:
  -h, --help    show this help message and exit
  --ask-kill    Ask before killing if there are multiple running services accessed by the token.
  --force-kill  Force kill the token's currently running services. Killing the token's services is enabled by default.
  --no-kill     Skip killing the token's currently running services.


$ waiter maintenance stop -h
usage: waiter maintenance stop [-h] [--no-ping | --ping] [--timeout TIMEOUT] token
...
  --no-ping             Skip pinging the token. Pinging the token is enabled by default.
  --ping                Ping the token after stopping maintenance.
  --timeout TIMEOUT, -t TIMEOUT
                        read timeout (in seconds) for ping request
```

### maintenance start

```
$ waiter --url 'http://localhost:9091' maintenance start test-token 'foo'
Successfully updated test-token.
Killing service w9091-servicegocjaybjge-7b9c39faa28efd0a7870207764751865 in http://localhost:9091...
Successfully killed w9091-servicegocjaybjge-7b9c39faa28efd0a7870207764751865 in http://localhost:9091.


$ waiter --url 'http://localhost:9091' maintenance start --no-kill test-token 'foo2'
Successfully updated test-token.


$ waiter --url 'http://localhost:9091' maintenance start --ask-kill test-token 'foo2'
No changes detected for test-token.
Killing service w9091-servicegocjaybjge-7b9c39faa28efd0a7870207764751865 in http://localhost:9091...
Successfully killed w9091-servicegocjaybjge-7b9c39faa28efd0a7870207764751865 in http://localhost:9091.


$ waiter --url 'http://localhost:9091' maintenance start --ask-kill test-token 'foo3'
Successfully updated test-token.
There are 2 services using token test-token:

=== http://localhost:9091 / w9091-servicegocjaybjge-e6c73b60d4c75ca0da89a4114b335226 ===
...
Kill service in http://localhost:9091? y
Killing service w9091-servicegocjaybjge-e6c73b60d4c75ca0da89a4114b335226 in http://localhost:9091...
Successfully killed w9091-servicegocjaybjge-e6c73b60d4c75ca0da89a4114b335226 in http://localhost:9091.

=== http://localhost:9091 / w9091-servicegocjaybjge-7b9c39faa28efd0a7870207764751865 ===
...
Kill service in http://localhost:9091? y
Killing service w9091-servicegocjaybjge-7b9c39faa28efd0a7870207764751865 in http://localhost:9091...
Successfully killed w9091-servicegocjaybjge-7b9c39faa28efd0a7870207764751865 in http://localhost:9091.


$ waiter --url 'http://localhost:9091' maintenance start --ask-kill test-token 'foo4'
Successfully updated test-token.
There are 2 services using token test-token:
=== http://localhost:9091 / w9091-servicegocjaybjge-d0e8ba3b23e5173d806d473f2b55d883 ===
...
Kill service in http://localhost:9091? n

=== http://localhost:9091 / w9091-servicegocjaybjge-e6c73b60d4c75ca0da89a4114b335226 ===
...
Kill service in http://localhost:9091? y
Killing service w9091-servicegocjaybjge-e6c73b60d4c75ca0da89a4114b335226 in http://localhost:9091...
Successfully killed w9091-servicegocjaybjge-e6c73b60d4c75ca0da89a4114b335226 in http://localhost:9091.
```

### maintenance stop

```
$ waiter --url 'http://localhost:9091' maintenance stop test-token 
Successfully updated test-token.
Pinging token test-token in http://localhost:9091...
Ping successful.
Service is currently Running.


$ waiter --url 'http://localhost:9091' maintenance stop --no-ping test-token 
Successfully updated test-token.
```

### CODE REVIEW

File: cli/waiter/subcommands/maintenance.py [hunk 1]
```diff
@@ -53,23 +57,62 @@ def check_maintenance(clusters, args, enforce_cluster):
...
     json_body.pop("maintenance")
-    return _update_token(cluster, token_name, existing_token_etag, json_body)
+    return_code, token_etag = _update_token(cluster, token_name, existing_token_etag, json_body)
+    if return_code == 0:
+        if ping_token:
+            if token_etag:
+                success = ping_token_on_cluster(cluster, token_name, timeout, wait_for_ping, token_etag)
+                return 0 if success else 1
+            else:
+                logging.debug(f'Not pinging token {token_name} in {cluster} as token ETag is missing.')
```
Comment (kevo1ution) [inline cli/waiter/subcommands/maintenance.py:110]: Should we return 1 in this case?

File: cli/waiter/subcommands/maintenance.py [hunk 2]
```diff
@@ -53,23 +57,62 @@ def check_maintenance(clusters, args, enforce_cluster):
...
+            force_flag = False
+        logging.debug(f"maintenance start will {'' if kill_services else 'not '} "
+                      f"{'force ' if force_flag else ''} kill services")
+        if kill_services:
+            if token_etag:
+                timeout_secs = 5
+                success = process_kill_request(clusters, token_name, False, force_flag, timeout_secs, True)
+                return 0 if success else 1
+            else:
+                logging.debug(f'Not killing services for token {token_name} in {cluster} as token ETag is missing.')
```
Comment (kevo1ution) [inline cli/waiter/subcommands/maintenance.py:83]: Should we return 1 in this case?

File: cli/waiter/subcommands/maintenance.py (hunk 1 above)
Comment (shamsimam) [reply cli/waiter/subcommands/maintenance.py:111]: Done.

File: cli/waiter/subcommands/maintenance.py (hunk 2 above)
Comment (shamsimam) [reply cli/waiter/subcommands/maintenance.py:83]: Done.

**Review by kevo1ution — APPROVED** [review]

**Pull request merged**